core/builder/src/main/java/io/quarkus/builder/BuildChainBuilder.java

```diff
@@ -34,9 +36,9 @@
 
     private final BuildStepBuilder finalStep;
     private final List<BuildProvider> providers = new ArrayList<>();
-    private final Map<BuildStepBuilder, StackTraceElement[]> steps = new HashMap<BuildStepBuilder, StackTraceElement[]>();
-    private final Set<ItemId> initialIds = new HashSet<>();
-    private final Set<ItemId> finalIds = new HashSet<>();
+    private final Map<BuildStepBuilder, StackTraceElement[]> steps = new LinkedHashMap<BuildStepBuilder, StackTraceElement[]>();
+    private final Set<ItemId> initialIds = new LinkedHashSet<>();
+    private final Set<ItemId> finalIds = new LinkedHashSet<>();
```

Comment:
I'm fairly certain these don't need to be linked in order to produce a reproducible order (since steps are already run concurrently).

Reply:
+1 Good catch, thanks!

Reply:
After some testing, `finalIds` turns out to require a LinkedHashSet, while initialIds does not.

Reply:
`finalIds` still a LinkedHashSet, initialIds back a HashSet in 351945e